src/annotator/util/test/buckets-test.js

```diff
@@ -0,0 +1,275 @@
+      findClosestOffscreenAnchor(fakeAnchors, 'down');
+      // It will disregard the anchor without the highlights and not try to
+      // assess its boundingRect
+      assert.equal(fakeGetBoundingClientRect.callCount, fakeAnchors.length - 1);
+    });
+
+    it('returns null if no valid anchor found', () => {
+      stubbedInnerHeight = sinon.stub(window, 'innerHeight').value(800);
+      assert.isNull(findClosestOffscreenAnchor([{ highlights: [] }], 'down'));
+      assert.isNull(findClosestOffscreenAnchor(fakeAnchors, 'down'));
+    });
+  });
+
+  describe('constructPositionPoints', () => {
+    let fakeAnchors;
+    let stubbedInnerHeight;
+
+    beforeEach(() => {
+      const fakeAnchor = fakeAnchorFactory();
+      fakeAnchors = [
+        fakeAnchor(), // top: 1
+        fakeAnchor(), // top: 101
+        fakeAnchor(), // top: 201
+        fakeAnchor(), // top: 301
+        fakeAnchor(), // top: 401
+        fakeAnchor(), // top: 501
+      ];
+      stubbedInnerHeight = sinon.stub(window, 'innerHeight').value(410);
+    });
+
+    afterEach(() => {
+      stubbedInnerHeight.restore();
+    });
+
+    it('returns an Array of anchors that are offscreen above', () => {
+      const positionPoints = constructPositionPoints(fakeAnchors);
+
+      assert.deepEqual(positionPoints.above, [fakeAnchors[0], fakeAnchors[1]]);
+    });
+
+    it('returns an Array of anchors that are offscreen below', () => {
+      const positionPoints = constructPositionPoints(fakeAnchors);
+
+      assert.deepEqual(positionPoints.below, [fakeAnchors[4], fakeAnchors[5]]);
+    });
+
+    it('does not return duplicate anchors', () => {
+      const positionPoints = constructPositionPoints([
+        fakeAnchors[0],
+        fakeAnchors[0],
+        fakeAnchors[5],
+        fakeAnchors[5],
+      ]);
+
+      assert.deepEqual(positionPoints.above, [fakeAnchors[0]]);
+      assert.deepEqual(positionPoints.below, [fakeAnchors[5]]);
+    });
+
+    it('returns an Array of position points for on-screen anchors', () => {
+      const positionPoints = constructPositionPoints(fakeAnchors);
+
+      // It should return two "point" positions for each on-screen anchor,
+      // one representing the top of the anchor's highlight box, one representing
+      // the bottom position
+      assert.equal(positionPoints.points.length, 4);
+      // The top position of the first on-screen anchor
+      assert.deepEqual(positionPoints.points[0], [201, 1, fakeAnchors[2]]);
+      // The bottom position of the first on-screen anchor
+      assert.deepEqual(positionPoints.points[1], [251, -1, fakeAnchors[2]]);
+      // The top position of the second on-screen anchor
+      assert.deepEqual(positionPoints.points[2], [301, 1, fakeAnchors[3]]);
+      // The bottom position of the second on-screen anchor
+      assert.deepEqual(positionPoints.points[3], [351, -1, fakeAnchors[3]]);
+    });
+
+    it('sorts on-screen points based on position type secondarily', () => {
```

Comment:
I found this test description hard to parse. It looks like you meant, "sorts points by position, then type"?

Reply:
Updated test description.